services/code_hosting/git_hub_service.rb

```diff
@@ -12,6 +17,148 @@ class GitHubService < CodeHostingService
 
     class << self
       attr_accessor :status_context_prefix
+
+      attr_writer :temporary_git_storage
+
+      attr_accessor :temporary_storage_path
+
+      attr_writer :cache
+
+      def cache
+        @cache ||= {}
+        return @cache
+      end
+
+      def temp_path
+        temporary_path = File.join(Dir.tmpdir, ".fastlane")
+        FileUtils.mkdir_p(temporary_path) unless File.directory?(temporary_path)
+        return temporary_path
+      end
+
+      def temporary_git_storage
+        temp_storage = File.join(Dir.tmpdir, ".tmp")
+        @temporary_git_storage ||= temp_storage
+        FileUtils.mkdir_p(@temporary_git_storage) unless File.directory?(@temporary_git_storage)
+        return @temporary_git_storage
+      end
+
+      # This method shallow clones the repo using the provider credential and looks for a valid _fastlane_
+      # configuration, parses the Fastfile (if found), and returns the lanes in the following way:
+      #     {
+      #       :ios => [:lane_name, :other_lane_name],
+      #       :android => [:just_other_lane],
+      #       :no_platform => [:a_not_platform_lane]
+      #     }
+      # @param repo_url [String]
+      # @param branch [String]
+      # @param provider_credential [GithubProviderCredential]
+      def peek_fastfile_configuration(repo_url: nil, branch: "master", provider_credential: nil, cache: true)
+        repo = repo_from_url(repo_url)
+        return self.cache[repo + branch] if cache && self.cache && !self.cache[repo + branch].nil? && self.cache[repo + branch].kind_of?(Hash)
```

Comment:
So I thought this could be a nice idea, an in-memory `repo` + `branch` cache, this is alive as long as the application exists, but I'm storing all `fastlane.json` from every peeked repo, so it can even be loaded between executions. The `cache` flag lets the user reload the whole thing just in case he made a recent change in the Fastfile.